src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs

```diff
@@ -451,35 +473,111 @@ private static bool IsSafeForReordering(BoundExpression expression, RefKind kind
+                    startIndex: 0,
+                    length: arguments.Length,
+                    argumentsAreVisited,
+                    rewrittenArguments: actualArguments,
+                    refKinds: refKinds,
+                    storesToTemps: storesToTemps,
+                    tempsBuilder: temporariesBuilder);
+            }
+            else
+            {
+                argumentsWillBeReused = true;
+                visitInterpolatedStringArguments(ref visitedReceiver, argumentRefKindsOpt);
+            }
 
             // all the formal arguments, except missing optionals, are now in place. 
-            // Optimize away unnecessary temporaries.
-            // Necessary temporaries have their store instructions merged into the appropriate 
-            // argument expression.
-            OptimizeTemporaries(actualArguments, storesToTemps, temporariesBuilder);
+            if (!argumentsWillBeReused)
+            {
+                // Optimize away unnecessary temporaries.
+                // Necessary temporaries have their store instructions merged into the appropriate
+                // argument expression.
+                OptimizeTemporaries(actualArguments, storesToTemps, temporariesBuilder);
+            }
 
             // Step two: If we have a params array, build the array and fill in the argument.
             if (expanded)
             {
-                actualArguments[actualArguments.Length - 1] = BuildParamsArray(syntax, methodOrIndexer, argsToParamsOpt, rewrittenArguments, parameters, actualArguments[actualArguments.Length - 1]);
+                var array = VisitAndBuildParamsArray(syntax, methodOrIndexer, argsToParamsOpt, arguments, parameters, actualArguments[actualArguments.Length - 1], argumentsAreVisited);
+
+                if (argumentsWillBeReused)
+                {
+                    BoundLocal boundTemp = _factory.StoreToTemp(array, out BoundAssignmentOperator store);
+                    actualArguments[^1] = boundTemp;
+                    storesToTemps.Add(store);
+                }
+                else
+                {
+                    actualArguments[^1] = array;
+                }
+            }
+
+
+            if (argumentsWillBeReused)
+            {
+                // We can't optimize away the temporaries if arguments are going to be reused,
+                // because we don't know whether a specific index is going to be reused or not
+                // (and therefore re-evaluated).
+                foreach (BoundExpression expression in storesToTemps)
+                {
+                    Debug.Assert(expression is BoundSequence or BoundAssignmentOperator);
+                    if (expression is BoundAssignmentOperator assignment)
+                    {
+                        temporariesBuilder.Add(((BoundLocal)assignment.Left).LocalSymbol);
```

Comment:
>temporariesBuilder.Add(((BoundLocal)assignment.Left).LocalSymbol); [](http://example.com/codeflow?start=24&length=66)

Why are we not adding temps as we allocate them?

Reply:
Will be entirely different with the 2-phase version.